// nest-next/dist/render.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderService = void 0;
const path_1 = __importDefault(require("path"));
const common_1 = require("@nestjs/common");
const next_utils_1 = require("./next-utils");
const route_regex_1 = require("./vendor/next/route-regex");
const interpolate_dynamic_path_1 = require("./vendor/next/interpolate-dynamic-path");
const is_dynamic_1 = require("./vendor/next/is-dynamic");
class RenderService {
    constructor() {
        this.initialized = false;
        this.config = {
            dev: process.env.NODE_ENV !== 'production',
            passthrough404: false,
            viewsDir: '/views',
        };
        this.dynamicRouteRegexes = new Map();
    }
    static init(config, handler, renderer, errorRenderer, server) {
        const self = new RenderService();
        self.mergeConfig(config);
        self.setRequestHandler(handler);
        self.setRenderer(renderer);
        self.setErrorRenderer(errorRenderer);
        self.bindHttpServer(server);
        return self;
    }
    mergeConfig(config) {
        var _a;
        if (typeof config.dev === 'boolean') {
            this.config.dev = config.dev;
        }
        if (typeof config.viewsDir === 'string' || config.viewsDir === null) {
            this.config.viewsDir = config.viewsDir;
        }
        if (typeof config.passthrough404 === 'boolean') {
            this.config.passthrough404 = config.passthrough404;
        }
        if (typeof config.basePath === 'string') {
            this.config.basePath = config.basePath;
        }
        if ((_a = config.dynamicRoutes) === null || _a === void 0 ? void 0 : _a.length) {
            this.initializeDynamicRouteRegexes(config.dynamicRoutes);
        }
    }
    setViewsDir(path) {
        this.config.viewsDir = path;
    }
    getViewsDir() {
        return this.config.viewsDir;
    }
    setIsDev(dev) {
        this.config.dev = dev;
    }
    isDev() {
        return this.config.dev;
    }
    passthroughNotFoundErrors() {
        return !!this.config.passthrough404;
    }
    setRequestHandler(handler) {
        this.requestHandler = handler;
    }
    getRequestHandler() {
        return this.requestHandler;
    }
    setRenderer(renderer) {
        this.renderer = renderer;
    }
    getRenderer() {
        return this.renderer;
    }
    setErrorRenderer(errorRenderer) {
        this.errorRenderer = errorRenderer;
    }
    getErrorRenderer() {
        return this.errorRenderer;
    }
    setErrorHandler(handler) {
        this.errorHandler = handler;
    }
    getErrorHandler() {
        return this.errorHandler;
    }
    isInternalUrl(url) {
        if (this.config.basePath && url.startsWith(this.config.basePath)) {
            return (0, next_utils_1.isInternalUrl)(url.replace(this.config.basePath, ''));
        }
        return (0, next_utils_1.isInternalUrl)(url);
    }
    initializeDynamicRouteRegexes(routes = []) {
        for (const route of routes) {
            const pathname = this.getNormalizedPath(route);
            this.dynamicRouteRegexes.set(route, (0, route_regex_1.getNamedRouteRegex)(pathname));
        }
    }
    isInitialized() {
        return this.initialized;
    }
    bindHttpServer(server) {
        if (this.initialized) {
            throw new Error('RenderService: already initialized');
        }
        this.initialized = true;
        const renderer = this.getRenderer();
        const getViewPath = this.getViewPath.bind(this);
        server.render = (response, view, data) => {
            const isFastify = response.request !== undefined;
            const res = isFastify ? response.res : response;
            const req = isFastify ? response.request.raw : response.req;
            if (req && res && renderer) {
                if (isFastify) {
                    response.sent = true;
                }
                return renderer(req, res, getViewPath(view, req.params), data);
            }
            else if (!renderer) {
                throw new common_1.InternalServerErrorException('RenderService: renderer is not set');
            }
            else if (!res) {
                throw new common_1.InternalServerErrorException('RenderService: could not get the response');
            }
            else if (!req) {
                throw new common_1.InternalServerErrorException('RenderService: could not get the request');
            }
            throw new Error('RenderService: failed to render');
        };
        let isFastifyAdapter = false;
        try {
            const { FastifyAdapter } = require('@nestjs/platform-fastify');
            isFastifyAdapter = server instanceof FastifyAdapter;
        }
        catch (e) {
        }
        if (isFastifyAdapter) {
            server
                .getInstance()
                .decorateReply('render', function (view, data) {
                const res = this.res;
                const req = this.request.raw;
                if (!renderer) {
                    throw new common_1.InternalServerErrorException('RenderService: renderer is not set');
                }
                this.sent = true;
                return renderer(req, res, getViewPath(view, req.params), data);
            });
        }
        else {
            server.getInstance().use((req, res, next) => {
                res.render = ((view, data) => {
                    if (!renderer) {
                        throw new common_1.InternalServerErrorException('RenderService: renderer is not set');
                    }
                    return renderer(req, res, getViewPath(view, req.params), data);
                });
                next();
            });
        }
    }
    getNormalizedPath(view) {
        var _a;
        const basePath = (_a = this.getViewsDir()) !== null && _a !== void 0 ? _a : '';
        const denormalizedPath = [basePath, view].join(view.startsWith('/') ? '' : '/');
        const pathname = path_1.default.posix.normalize(denormalizedPath);
        return pathname;
    }
    getViewPath(view, params) {
        const pathname = this.getNormalizedPath(view);
        if (!(0, is_dynamic_1.isDynamicRoute)(pathname)) {
            return pathname;
        }
        const regex = this.dynamicRouteRegexes.get(pathname);
        if (!regex) {
            console.warn(`RenderService: view ${view} is dynamic and has no route regex`);
        }
        return (0, interpolate_dynamic_path_1.interpolateDynamicPath)(pathname, params, regex);
    }
}
exports.RenderService = RenderService;
